pkg/grpc/labelstore/client/client.go

```diff
@@ -51,15 +59,31 @@ func (c Client) WatchMatches(selector klabels.Selector, labelType labels.Type, q
 		for {
 			labeled, err := watchClient.Recv()
 			if grpc.Code(err) == codes.Canceled {
+				c.logger.Infoln("label store client: terminating WatchMatches()")
 				// This just means quitCh fired and the RPC was canceled as expected
 				return
 			}
 
 			if err != nil {
-				// doesn't hurt to call cancelFunc() more than once potentially
-				cancelFunc()
-				c.logger.WithError(err).Errorln("Unexpected error reading from WatchMatches stream")
-				return
+				innerCancel()
+				c.logger.WithError(err).Errorln("unexpected error reading from WatchMatches stream, starting another RPC")
+
+				watchClient = nil
+
+				for watchClient == nil {
+					innerCtx, innerCancel = context.WithCancel(outerCtx)
+
+					time.Sleep(2 * time.Second)
```

Comment:
Is there a best practice guide from G here? I'd expect to see an exponential vs linear backoff.

I also wonder if we need a global timeout.